python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -423,6 +438,35 @@ class DbtCliInvocation:
     _stdout: List[Union[str, Dict[str, Any]]] = field(init=False, default_factory=list)
     _error_messages: List[str] = field(init=False, default_factory=list)
 
+    # Caches fetching relation column metadata to avoid redundant queries to the database.
+    _relation_column_metadata_cache: Dict[RelationKey, RelationData] = field(
+        init=False, default_factory=dict
+    )
+
+    def _get_columns_from_dbt_resource_props(
+        self, adapter: BaseAdapter, dbt_resource_props: Dict[str, Any]
+    ) -> RelationData:
+        """Given a dbt resource properties dictionary, fetches the resource's column metadata from
+        the database, or returns the cached metadata if it has already been fetched.
+        """
+        relation_key = RelationKey(
+            database=dbt_resource_props["database"],
+            schema=dbt_resource_props["schema"],
+            identifier=(
+                dbt_resource_props["identifier"]
+                if dbt_resource_props["unique_id"].startswith("source")
+                else dbt_resource_props["alias"]
+            ),
+        )
+        if relation_key in self._relation_column_metadata_cache:
+            return self._relation_column_metadata_cache[relation_key]
+
+        relation = _get_relation_from_adapter(adapter=adapter, relation_key=relation_key)
+        cols: List = adapter.get_columns_in_relation(relation=relation)
+        data = RelationData(name=str(relation), columns=cols)
+        self._relation_column_metadata_cache[relation_key] = data
```

Comment:
This should be in a separate method. Also, since multiple threads are accessing this, should there be some locking in place?

Reply:
Dict access should be thread-safe, with the caveat that without a lock, multiple writes are allowed to happen (if both pass the `is key already present` gate before a write happens), and would be sequenced arbitrarily. In this case, that should be fine, since the stored data will be identical - it just means data will be fetched twice in rare scenarios. This prioritizes throughput over minimizing number of adapter accesses, which I think is preferable vs locking other threads each time there's a write.